From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Mon, 21 Aug 2023 21:55:13 +0800
Subject: [PATCH] Setup project and rebrand


diff --git a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
index 80b8a2483d9fa401be1370e76e902d735df6c540..96ae0a53c3dcddb8c33f20f36157074b8541061e 100644
--- a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
+++ b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
@@ -21,7 +21,7 @@ public class PaperVersionFetcher implements VersionFetcher {
     private static final java.util.regex.Pattern VER_PATTERN = java.util.regex.Pattern.compile("^([0-9\\.]*)\\-.*R"); // R is an anchor, will always give '-R' at end
     private static final String GITHUB_BRANCH_NAME = "master";
     // Kaiiju start
-    private static final String DOWNLOAD_PAGE = "https://github.com/Era4FunMC/Molia"; //Molia
+    private static final String DOWNLOAD_PAGE = "https://github.com/Era4FunMC/DirtyMolia"; //Molia
     private static int distance = -2; public int distance() { return distance; }
     // Kaiiju end
     private static @Nullable String mcVer;
@@ -33,8 +33,8 @@ public class PaperVersionFetcher implements VersionFetcher {
     @Nonnull
     @Override
     public Component getVersionMessage(@Nonnull String serverVersion) {
-        String[] parts = serverVersion.substring("git-Molia-".length()).split("[-\\s]"); // Folia // Kaiiju //Molia
-        final Component updateMessage = getUpdateStatusMessage("Era4FunMC/Molia", "ver/" + getMinecraftVersion(), parts[0]); // Folia // Kaiiju //Molia
+        String[] parts = serverVersion.substring("git-DirtyMolia-".length()).split("[-\\s]"); // Folia // Kaiiju //Molia
+        final Component updateMessage = getUpdateStatusMessage("Era4FunMC/DirtyMolia", "ver/" + getMinecraftVersion(), parts[0]); // Folia // Kaiiju //Molia
         final Component history = getHistory();
 
         return history != null ? TextComponent.ofChildren(updateMessage, Component.newline(), history) : updateMessage;
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 05dfe6b7266346f59d846269fc3f1aa49aa95981..a9de0e7ddbc9367a992b98be56d9b2b86d946978 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1816,7 +1816,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return "Molia"; //Molia - Molia > //Kaiiju - Kaiiju > // Folia - Folia > // Paper - Paper > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
+        return "DirtyMolia"; //Molia - Molia > //Kaiiju - Kaiiju > // Folia - Folia > // Paper - Paper > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
     }
 
     public SystemReport fillSystemReport(SystemReport details) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index a3db94d40029060739ca24d746ba740b18e86104..fed7cd914dc2a035110c35614fee92a0d07d5a69 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -266,7 +266,7 @@ import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
 
 public final class CraftServer implements Server {
-    private final String serverName = "Molia"; // Kaiiju // Folia // Paper //Molia
+    private final String serverName = "DirtyMolia"; // Kaiiju // Folia // Paper //Molia
     private final String serverVersion;
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
