From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Sat, 26 Aug 2023 13:44:17 +0800
Subject: [PATCH] Change default server mod name to DirtyMolia


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index 051c807211c369b31442917d35cbe5c09681d95a..d8b41a0035741214ff19357dc4598b09b95e8d80 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -37,7 +37,7 @@ public class MoliaConfig {
     public static boolean fakeplayerResident = false;
     public static List<String> unableFakeplayerNames = Lists.newArrayList();
     public static int fakeplayerLimit = 10;
-    public static String serverModName = "Molia";
+    public static String serverModName = "DirtyMolia";
     //
 
     private static final YamlConfiguration configEntry = new YamlConfiguration();
