From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: M2ke4U <79621885+MrHua269@users.noreply.github.com>
Date: Fri, 3 Nov 2023 22:00:35 +0800
Subject: [PATCH] Fix forgetting updating worlddata


diff --git a/src/main/java/io/papermc/paper/threadedregions/RegionizedData.java b/src/main/java/io/papermc/paper/threadedregions/RegionizedData.java
index 79ffa26d86efff3c5be3b56330c0d0e9f37f0fc7..5fe2b5cbb7410dfb12f10ead6f21437269b82c6c 100644
--- a/src/main/java/io/papermc/paper/threadedregions/RegionizedData.java
+++ b/src/main/java/io/papermc/paper/threadedregions/RegionizedData.java
@@ -154,7 +154,13 @@ public final class RegionizedData<T> {
             throw new IllegalStateException("World check failed: expected world: " + this.world.getWorld().getKey() + ", region world: " + region.getData().world.getWorld().getKey());
         }
 
-        return region.getData().getOrCreateRegionizedData(this);
+        final T ret = region.getData().getOrCreateRegionizedData(this);
+
+        if (ret instanceof RegionizedWorldData data){
+            Level.WORLD_DATA_POOL.updateRegion(region,data);
+        }
+
+        return ret;
     }
 
     /**
diff --git a/src/main/java/io/papermc/paper/threadedregions/TickRegions.java b/src/main/java/io/papermc/paper/threadedregions/TickRegions.java
index f5cc85bafed810722d11adc7b53f92c7d6b8808d..d024827753e8d68badaa0b2323d2e1ccb44f5a5d 100644
--- a/src/main/java/io/papermc/paper/threadedregions/TickRegions.java
+++ b/src/main/java/io/papermc/paper/threadedregions/TickRegions.java
@@ -17,6 +17,7 @@ import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.chunk.LevelChunk;
 import org.bukkit.Bukkit;
+import org.bukkit.World;
 import org.slf4j.Logger;
 import java.util.Iterator;
 import java.util.concurrent.TimeUnit;
