From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Sat, 12 Aug 2023 08:24:13 +0800
Subject: [PATCH] Fix rebrand


diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index f6ed9ec83c825c163be790afe6410aefeef2a4b8..43ac057cc5a2529d6c5f55c8c9988b0debd3630f 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -11,7 +11,7 @@ public final class Versioning {
     public static String getBukkitVersion() {
         String result = "Unknown-Version";
 
-        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/me.earthme.molia/molia-api/kaiiju-api/pom.properties"); // Folia // Kaiiju // Folia //Molia
+        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/me.earthme.molia/molia-api/pom.properties"); // Folia // Kaiiju // Folia //Molia
         Properties properties = new Properties();
 
         if (stream != null) {
