From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Tue, 29 Aug 2023 12:52:41 +0800
Subject: [PATCH] Remove debug methods in FallingBlockEntity


diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index 60a5d9aa94d1da65130aa736245b9efdfa7fffac..bf763869640298fc2e2fcd9b8d832d097f747683 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -246,8 +246,6 @@ public class FallingBlockEntity extends Entity {
                                     this.spawnAtLocation((ItemLike) block);
                                 }
                             } else {
-                                System.out.println("F1 + "+flag1 +" F2 " + flag2 + " F3 "+flag3);
-                                System.out.println(this.level().getBlockState(blockposition.below()).getBlock().getName());
                                 this.discard();
                                 if (this.dropItem && this.level().getGameRules().getBoolean(GameRules.RULE_DOENTITYDROPS)) {
                                     this.callOnBrokenAfterFall(block, blockposition);
